[concurrency][distributed] Inject invocations of transport.actorReady in init

kavon opened the pull request:
A distributed actor's `init` and `deinit` should invoke the ActorTransport's `actorReady` and `resignIdentity` functions, respectively, in order to notify the transport of the lifetimes of distributed actor that are bound to the transport. We're currently missing the calls to `actorReady` (and for failable inits, the `resignIdentity`), and this PR fixes that.

Resolves: rdar://81783599

kavon committed:
fixed emission of actorReady call in init epilogue

kavon committed:
impose `self` restrictions in distributed actor inits

I initially left-out these restrictions, while
the design of distributed actor inits was in flux.
Now that the dust is somewhat settled, we need the same
restrictions on `self` that regular actors have, because
we do not want the local instance to have races in the init.

kavon committed:
refactor some of SILGen's distributed actor code

kavon committed:
invoke resignIdentity in an async failable initializer

For distributed actors, their async initializers will call
actorReady prior to the end of the initializer.
If that happens, we need to resign the identity if we end
up in the failure path of the init.

ktoso reviewed (commented):
Looking good so far :)

kavon committed:
decouple the emission of actorReady from SILGen

We need to be able to inject a call to a distributed actor's
transport.actorReady, passing the actor instance to it,
during definite initialization. This means that its dependence
on SILGenFunction must be broken, hence this refactoring as
a SILOptimizer utility.

kavon committed:
inject actorReady calls for async dist actor ctors

Immediately after the hop_to_executor in an async, distributed
actor init, we need to notify the transport that the actor is ready.

This patch does not yet account for more complex cases. In particular,
we will need a mechanism to prevent multiple calls to actorReady,
which can happen if a loop appears in the init:

distributed actor Dactor {
  var x: Int
  init(tport: ActorTransport, count: Int) async {
    var i = count
    repeat {
      self.x = count
      // hop is injected here
      i -= 1
    } while i > 0
  }
}

ktoso commented:
@swift-ci please smoke test

kavon committed:
update dist actor deinit SIL tests

I've combined the coverage of two tests
into one file.

kavon committed:
merge coverage of multiple transport args into a runtime test

the SIL test for ensuring that the transport argument is chosen correctly can
also be covered by a runtime test. It also gives me an excuse to add some
async init coverage too.

kavon committed:
Move the few SIL tests under the Distributed directory

Basically turn test/SIL/Distributed into test/Distributed/SIL

Since we have typechecking and runtime tests under there already,
and there are only a handful of distributed-actor specific SIL tests.

kavon committed:
add SIL coverage for current actorReady and resignIdentity calls

kavon committed:
Replace `destoryDistributedActor` builtin with `destroyDefaultActor`

Currently, they both end up doing the exact same thing, so there's
no need for both.

kavon committed:
remove redundant defers, now that actorReady is called implicitly

kavon committed:
capture the known problem with multiple actorReady calls in async inits

We haven't yet integrated more deeply into DI to prevent the multiple calls.

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -315,163 +240,37 @@ static void emitDistributedActorStore_init_assignIdentity(
+  selfArg = selfArg.borrow(*this, loc);
+  emitTransportInit(*this, ctor, loc, selfArg);
+  emitIdentityInit(*this, ctor, loc, selfArg);
 }
```
thanks for the great cleanups here by the way 👍

ktoso commented on lib/SILOptimizer/Transforms/AccessEnforcementReleaseSinking.cpp:
```diff
@@ -182,7 +182,6 @@ static bool isBarrier(SILInstruction *inst) {
     case BuiltinValueKind::InitializeDefaultActor:
     case BuiltinValueKind::DestroyDefaultActor:
     case BuiltinValueKind::InitializeDistributedRemoteActor:
-    case BuiltinValueKind::DestroyDistributedActor:
```
yep, good to remove -- thank you

ktoso commented on test/Distributed/SIL/distributed_actor_default_init_sil.swift:
```diff
@@ -0,0 +1,131 @@
+
+}
+
+
```
Thanks a lot for those tests -- they're annoying to write heh... :)

kavon committed:
throwing sync inits are also not fully working

kavon committed:
fix whitespace in check line for Linux
